bestofjs/bestofjs
# Prepare Rising Stars 2024

## Goal

It's time to prepare the _Rising Stars 2024_!

This PR migrates the script to generate the data consumed by the application, from https://github.com/bestofjs/bestofjs-backend to this monorepo.

Two tasks are added to the backend app:

- `build-rising-stars`: scan all projects and generate JSON data, based on the `categories` defined in Rising Stars.
- `fetch-missing-snapshots`: call GitHub API to fetch stargazers data. It's useful for some popular projects that were added to _Best of JS_ during the year, when we miss data before projects were added. The script will add one snapshot by month, to be able to show the monthly graph.

Added dependencies:

- [Oktokit](https://github.com/octokit/octokit.js) used to handle pagination when calling GitHub API
- [node-fetch-cache](https://github.com/mistval/node-fetch-cache) used to cache data in the local file system, to be able to call multiple the GitHub API, it's very useful when debugging or just when running the script in `--dryMode` first

## How to test

Requirement: https://github.com/bestofjs/javascript-risingstars should be cloned in the parent folder of the monorepo.

In the future, it would make more sense to move Rising Stars to the monorepo.

This script will generate `risingstars.json` file in `apps/backend/build` folder

```sh
bun run apps/backend/src/cli.ts build-rising-stars --year 2024 --concurrency 5
```

To add the missing snapshots, targeting a single project by its `slug`:

```sh
bun run apps/backend/src/cli.ts fetch-missing-snapshots --slug bolt --dryRun
```

## Example of projects with missing snapshots

Running `fetch-missing-snapshots` added data required to display the graph of the monthly trends in https://bestofjs.org/projects/webvm

![image](https://github.com/user-attachments/assets/b710dffa-38f2-43fd-b498-b7c9a8368c87)
